src/charm.py

```diff
@@ -1,104 +1,400 @@
-        container = event.workload
-        # Define an initial Pebble layer configuration
-        pebble_layer = {
-            "summary": "httpbin layer",
-            "description": "pebble config layer for httpbin",
+        users = self._get_users_from_charm_config(users={})
+        self._push_container_config(users)
+
+    def _on_config_changed(self, _) -> None:
+        """Handle changes in configuration."""
+        # Handle any user changes that may have been created through config.
+        users = self._get_userlist_from_container()
+        users = self._get_users_from_charm_config(users=users)
+        self._push_container_config(users)
+
+        container = self.unit.get_container(self._pgbouncer_service)
+        layer = self._pgbouncer_layer()
+
+        if not container.can_connect():
+            self.unit.status = WaitingStatus("waiting for pebble in pgbouncer workload container.")
+            return
+
+        services = container.get_plan().to_dict().get("services", {})
+        if services != layer["services"]:
+            container.add_layer(self._pgbouncer_service, layer, combine=True)
+            logging.info("Added layer 'pgbouncer' to pebble plan")
+            container.restart(self._pgbouncer_service)
+            logging.info("restarted pgbouncer service")
+        self.unit.status = ActiveStatus()
+
+    def _pgbouncer_layer(self) -> Layer:
+        """Returns a default pebble config layer for the pgbouncer container."""
+        return {
+            "summary": "pgbouncer layer",
+            "description": "pebble config layer for pgbouncer",
             "services": {
-                "httpbin": {
-                    "override": "replace",
-                    "summary": "httpbin",
-                    "command": "gunicorn -b 0.0.0.0:80 httpbin:app -k gevent",
+                self._pgbouncer_service: {
+                    "summary": "pgbouncer service",
+                    "user": self._pgbouncer_user,
+                    "command": f"pgbouncer {INI_PATH}",
                     "startup": "enabled",
-                    "environment": {"thing": self.model.config["thing"]},
+                    "override": "replace",
+                    "environment": {
+                        "PGB_DATABASES": self.config["pgb_databases"],
+                        "PGB_LISTEN_PORT": self.config["pgb_listen_port"],
+                        "PGB_LISTEN_ADDRESS": self.config["pgb_listen_address"],
+                        "PGB_ADMIN_USERS": self.config["pgb_admin_users"],
+                    },
                 }
             },
         }
-        # Add intial Pebble config layer using the Pebble API
-        container.add_layer("httpbin", pebble_layer, combine=True)
-        # Autostart any services that were defined with startup: enabled
+
+    def _on_pgbouncer_pebble_ready(self, event) -> None:
+        """Define and start pgbouncer workload."""
```

Comment:
The pebble ready event is one place where you can probably get away with out it - the event is only fired precisely because pebble is up and juju can talk to it - now theoretically nothing stops it falling over before the hook executes, so you could choose to go belt and braces and check the connection